src/porepy/grids/grid.py

```diff
@@ -880,41 +881,123 @@ def update_boundary_node_tag(self) -> None:
                 ]
                 self.tags[node_tag][nodes] = True
 
-    def cell_diameters(self, cn: Optional[sps.spmatrix] = None) -> np.ndarray:
+    @lru_cache
+    def cell_diameters(
+        self,
+        cn: Optional[sps.spmatrix] = None,
+        cell_wise: bool = True,
+        func: Optional[Callable] = None,
+    ) -> np.ndarray:
         """Computes the cell diameters.
 
         Parameters:
             cn: ``default=None``
-                Cell-to-nodes map, already computed previously.
-                If None, a call to :meth:`cell_nodes` is provided.
+                Cell-to-nodes map, already computed previously. If None, a call to
+                :meth:`cell_nodes` is provided.
+
+            cell_wise:
+                If True, returns the cell diameters. If False, the parameter ``func``
+                must be provided to specify the aggregation function.
+
+                The requirement that ``func`` is provided is linked to technical
+                limitations of the implementation.
+
+            func:
+                A function to apply to the computed diameters. Must be provided if
+                ``cell_wise`` is False.
+
+        Raises:
+            ValueError: If ``func`` is not provided when ``cell_wise`` is False.
 
         Returns:
-            Values of the cell diameter for each cell, ``(shape=(num_cells))``.
+
+            If ``cell_wise`` is True, the cell diameter for each cell,
+            ``(shape=(num_cells))``. If ``cell_wise`` is False, the maximum diameter of
+            all cells.
 
             If the dimension of the grid is zero, returns 0.
 
         """
         if self.dim == 0:
             return np.zeros(1)
 
+        if cell_wise is False and func is None:
+            raise ValueError("func must be provided if cell_wise is False")
+
+        if cn is None:
+            cn = self.cell_nodes()
+
         def comb(n):
+            # Helper function to get all combinations of two elements in n.
             return np.fromiter(
```

Comment:
Could this benefit from an explanation?

Reply:
I don't know who wrote it (not me), but it is pretty clear to me what happens based on some knowledge of itertools. Also, I don't see us debugging this ever.